File: rudolph/src/main/java/cn/wzbos/android/rudolph/utils/TypeUtils.java
package cn.wzbos.android.rudolph.utils;

import android.content.Context;


import cn.wzbos.android.rudolph.RLog;
import cn.wzbos.android.rudolph.router.RouteBuilder;


/**
 * Created by wuzongbo on 2017/6/21.
 */
public class TypeUtils {

    private static int toInteger(String value) {
        int val = 0;
        if (!android.text.TextUtils.isEmpty(value)) {
            if (value.toLowerCase().startsWith("0x")) {
                val = Integer.valueOf(value.substring(2), 16);
            } else {
                val = Integer.parseInt(value);
            }
        }
        return val;
    }

    private static boolean toBoolean(String value) {
        boolean val = false;
        if (!android.text.TextUtils.isEmpty(value)) {
            if (value.toLowerCase().equals("true") || value.toLowerCase().equals("false")) {
                val = Boolean.parseBoolean(value);
            } else {
                val = Integer.parseInt(value) > 0;
            }
        }
        return val;
    }


    private static double toDouble(String value) {
        double val = 0;
        if (!android.text.TextUtils.isEmpty(value))
            val = Double.parseDouble(value);
        return val;
    }

    private static float toFloat(String value) {
        float val = 0;
        if (!android.text.TextUtils.isEmpty(value))
            val = Float.parseFloat(value);

        return val;
    }

    private static long toLong(String value) {
        long val = 0;
        if (!android.text.TextUtils.isEmpty(value))
            val = Long.parseLong(value);
        return val;
    }

    private static short toShort(String value) {
        short val = 0;
        if (!android.text.TextUtils.isEmpty(value)) {
            if (value.toLowerCase().startsWith("0x")) {
                val = Short.valueOf(value.substring(2), 16);
            } else {
                val = Short.parseShort(value);
            }
        }
        return val;
    }

    private static byte toByte(String value) {
        byte val = 0;
        if (!android.text.TextUtils.isEmpty(value)) {
            if (value.toLowerCase().startsWith("0x")) {
                val = Byte.valueOf(value.substring(2), 16);
            } else {
                val = Byte.parseByte(value);
            }
        }
        return val;
    }

    private static char toChar(String value) {
        char val = 0;
        if (!android.text.TextUtils.isEmpty(value))
            val = value.charAt(0);
        return val;
    }

    public static Object getObject(Context context, String name, String value, String target) {
        return getObject(context, name, value, target, null);
    }

    public static Object getObject(Context context, String name, String value, String target, RouteBuilder<?, ?> builder) {
        RLog.d("TypeUtils", name + " = " + value + ", target:" + target + ", bundle:" + builder);

        if ("java.lang.String".equals(target) || "java.lang.CharSequence".equals(target)) {  //String、CharSequence
            if (builder != null)
                builder.putExtra(name, value);
            return value;
        } else if ("int".equals(target) || "java.lang.Integer".equals(target)) {    //int、Integer
            int val = toInteger(value);
            if (builder != null)
                builder.putExtra(name, val);
            return val;
        } else if ("boolean".equals(target) || "java.lang.Boolean".equals(target)) {//boolean、Boolean
            boolean val = toBoolean(value);
            if (builder != null)
                builder.putExtra(name, val);
            return val;
        } else if ("double".equals(target) || "java.lang.Double".equals(target)) {  //double、Double
            double val = toDouble(value);
            if (builder != null)
                builder.putExtra(name, val);
            return val;
        } else if ("float".equals(target) || "java.lang.Float".equals(target)) {    //float、Float
            float val = toFloat(value);
            if (builder != null)
                builder.putExtra(name, val);
            return val;
        } else if ("long".equals(target) || "java.lang.Long".equals(target)) {      //long、Long
            long val = toLong(value);
            if (builder != null)
                builder.putExtra(name, val);
            return val;
        } else if ("short".equals(target) || "java.lang.Short".equals(target)) {    //short、Short
            short val = toShort(value);
            if (builder != null)
                builder.putExtra(name, val);
            return val;
        } else if ("byte".equals(target) || "java.lang.Byte".equals(target)) {      //byte、Byte
            byte val = toByte(value);
            if (builder != null)
                builder.putExtra(name, val);
            return val;
        } else if ("char".equals(target) || "java.lang.Character".equals(target)) { //char、Character
            char val = toChar(value);
            if (builder != null)
                builder.putExtra(name, val);
            return val;
        } else if ("android.content.Context".equals(target)) {                         //Context
            return context;
        } else {                                                    //Object
            if (builder != null)
                builder.putExtra(name, value);
            return value;
        }
    }
}

File: rudolph/src/main/java/cn/wzbos/android/rudolph/utils/TypeUtils.kt
package cn.wzbos.android.rudolph.utils

import android.app.Application
import android.content.Context
import android.util.Base64
import cn.wzbos.android.rudolph.ExtraInfo
import cn.wzbos.android.rudolph.Rudolph
import cn.wzbos.android.rudolph.logger.RLog
import cn.wzbos.android.rudolph.router.RouteBuilder
import com.google.gson.Gson

/**
 * Created by wuzongbo on 2017/6/21.
 */
object TypeUtils {
    private fun toInteger(value: String?, default: Int? = null): Int? {
        var ret = default
        if (!value.isNullOrEmpty()) {
            ret = if (value.startsWith("0x", true)) {
                Integer.valueOf(value.substring(2), 16)
            } else {
                value.toInt()
            }
        }
        return ret
    }

    private fun toBoolean(value: String?, default: Boolean? = false): Boolean? {
        var ret = default
        if (!value.isNullOrEmpty()) {
            ret =
                if (value.equals("true", true) || value.equals("false", true)) {
                    java.lang.Boolean.parseBoolean(value)
                } else {
                    value.toInt() > 0
                }
        }
        return ret
    }

    private fun toDouble(value: String?, default: Double? = 0.0): Double? {
        var ret = default
        if (!value.isNullOrEmpty()) ret = value.toDouble()
        return ret
    }

    private fun toFloat(value: String?, default: Float? = 0f): Float? {
        var ret = default
        if (!value.isNullOrEmpty()) ret = value.toFloat()
        return ret
    }

    private fun toLong(value: String?, default: Long? = 0): Long? {
        var ret = default
        if (!value.isNullOrEmpty()) ret = value.toLong()
        return ret
    }

    private fun toShort(value: String?, default: Short? = 0): Short? {
        var ret = default
        if (!value.isNullOrEmpty()) {
            ret = if (value.startsWith("0x", true)) {
                value.substring(2).toShort(16)
            } else {
                value.toShort()
            }
        }
        return ret
    }

    private fun toByte(value: String?, default: Byte? = 0): Byte? {
        var ret = default
        if (!value.isNullOrEmpty()) {
            ret = if (value.startsWith("0x", true)) {
                java.lang.Byte.valueOf(value.substring(2), 16)
            } else {
                value.toByte()
            }
        }
        return ret
    }

    private fun toChar(value: String?, default: Char? = Char.MIN_VALUE): Char? {
        var ret = default
        if (!value.isNullOrEmpty()) ret = value[0]
        return ret
    }

    fun getObject(context: Context?, name: String, value: String?, type: ExtraInfo): Any? {
        return getObject(context, name, value, type, null)
    }

    fun getObject(
        context: Context?,
        name: String,
        value: String?,
        extraType: ExtraInfo,
        builder: RouteBuilder<*, *>?,
    ): Any? {
        RLog.d("TypeUtils", "$name = $value, type:$extraType, bundle:$builder")
        return when (extraType.type) {
            String::class.java, CharSequence::class.java -> value?.also {
                builder?.putExtra(
                    name,
                    it
                )
            }
            Int::class.java -> toInteger(value, 0)?.also { builder?.putExtra(name, it) }
            Boolean::class.java -> toBoolean(value, false)?.also { builder?.putExtra(name, it) }
            Double::class.java -> toDouble(value, 0.0)?.also { builder?.putExtra(name, it) }
            Float::class.java -> toFloat(value, 0f)?.also { builder?.putExtra(name, it) }
            Long::class.java -> toLong(value, 0L)?.also { builder?.putExtra(name, it) }
            Short::class.java -> toShort(value)?.also { builder?.putExtra(name, it) }
            Byte::class.java -> toByte(value)?.also { builder?.putExtra(name, it) }
            Char::class.java -> toChar(value)?.also { builder?.putExtra(name, it) }
            Integer::class.java -> toInteger(value)?.also { builder?.putExtra(name, it) }
            java.lang.Boolean::class.java -> toBoolean(value)?.also { builder?.putExtra(name, it) }
            java.lang.Double::class.java -> toDouble(value)?.also { builder?.putExtra(name, it) }
            java.lang.Float::class.java -> toFloat(value)?.also { builder?.putExtra(name, it) }
            java.lang.Long::class.java -> toLong(value)?.also { builder?.putExtra(name, it) }
            java.lang.Short::class.java -> toShort(value)?.also { builder?.putExtra(name, it) }
            java.lang.Byte::class.java -> toByte(value)?.also { builder?.putExtra(name, it) }
            java.lang.Character::class.java -> toChar(value)?.also { builder?.putExtra(name, it) }
            Application::class.java -> Rudolph.context
            Context::class.java -> context ?: Rudolph.context
            else -> {
                var strValue = value
                var ret: Any? = null
                if (extraType.base64) {
                    strValue = String(Base64.decode(strValue, Base64.NO_PADDING or Base64.URL_SAFE))
                }
                if (extraType.json) {
                    ret = Gson().toJson(strValue, extraType.type)
                }
                //extra 应为原始值
                builder?.putExtra(name, value)
                ret
            }
        }
    }
}